File: android/src/main/java/com/markdown/styles/StyleConfig.kt
package com.markdown.styles

import android.content.Context
import android.graphics.Color
import android.graphics.Typeface
import org.json.JSONObject

data class ElementStyle(
    // Text
    val fontSize: Float = 0f,
    val fontWeight: String? = null,
    val fontStyle: String? = null,
    val fontFamily: String? = null,
    val lineHeight: Float = 0f,
    val letterSpacing: Float = 0f,
    val textDecorationLine: String? = null,
    val textDecorationColor: Int? = null,
    val textDecorationStyle: String? = null,
    val textAlign: String? = null,
    val color: Int? = null,

    // View (container)
    val backgroundColor: Int? = null,

    // Layout
    val gap: Float = 0f,
    val width: Float = 0f,
    val height: Float = 0f,
    val maxWidth: Float = 0f,
    val maxHeight: Float = 0f,

    // Padding
    val padding: Float = 0f,
    val paddingHorizontal: Float = 0f,
    val paddingVertical: Float = 0f,
    val paddingTop: Float = 0f,
    val paddingBottom: Float = 0f,
    val paddingLeft: Float = 0f,
    val paddingRight: Float = 0f,

    // Margin
    val margin: Float = 0f,
    val marginHorizontal: Float = 0f,
    val marginVertical: Float = 0f,
    val marginTop: Float = 0f,
    val marginBottom: Float = 0f,
    val marginLeft: Float = 0f,
    val marginRight: Float = 0f,

    // Border
    val borderColor: Int? = null,
    val borderWidth: Float = 0f,
    val borderRadius: Float = 0f,
    val borderLeftColor: Int? = null,
    val borderLeftWidth: Float = 0f,
    val borderRightColor: Int? = null,
    val borderRightWidth: Float = 0f,
    val borderTopColor: Int? = null,
    val borderTopWidth: Float = 0f,
    val borderBottomColor: Int? = null,
    val borderBottomWidth: Float = 0f,
) {
    fun resolveTypeface(): Typeface {
        val base = when (fontFamily) {
            "monospace", "Menlo" -> Typeface.MONOSPACE
            "serif" -> Typeface.SERIF
            else -> Typeface.DEFAULT
        }

        val style = when {
            fontWeight == "bold" && fontStyle == "italic" -> Typeface.BOLD_ITALIC
            fontWeight == "bold" || fontWeight == "600" || fontWeight == "700" -> Typeface.BOLD
            fontStyle == "italic" -> Typeface.ITALIC
            else -> Typeface.NORMAL
        }

        return Typeface.create(base, style)
    }

    /** Resolve typeface cascading over a base typeface. */
    fun resolveTypefaceWithBase(base: Typeface?): Typeface {
        val family = when (fontFamily) {
            "monospace", "Menlo" -> Typeface.MONOSPACE
            "serif" -> Typeface.SERIF
            null, "" -> base ?: Typeface.DEFAULT
            else -> base ?: Typeface.DEFAULT
        }

        val isBold = fontWeight == "bold" || fontWeight == "600" || fontWeight == "700"
        val isItalic = fontStyle == "italic"

        // If neither weight nor style is set, preserve the base style
        if (!isBold && !isItalic && fontWeight == null && fontStyle == null) {
            return if (fontFamily != null && fontFamily.isNotEmpty()) family else (base ?: Typeface.DEFAULT)
        }

        val style = when {
            isBold && isItalic -> Typeface.BOLD_ITALIC
            isBold -> Typeface.BOLD
            isItalic -> Typeface.ITALIC
            else -> Typeface.NORMAL
        }

        return Typeface.create(family, style)
    }

    fun resolvedFontSize(): Float = if (fontSize > 0) fontSize else 16f

    // --- Padding resolution (cascade: specific > directional > general) ---

    fun resolvedPaddingTop(): Float = when {
        paddingTop > 0f -> paddingTop
        paddingVertical > 0f -> paddingVertical
        padding > 0f -> padding
        else -> 0f
    }

    fun resolvedPaddingBottom(): Float = when {
        paddingBottom > 0f -> paddingBottom
        paddingVertical > 0f -> paddingVertical
        padding > 0f -> padding
        else -> 0f
    }

    fun resolvedPaddingLeft(): Float = when {
        paddingLeft > 0f -> paddingLeft
        paddingHorizontal > 0f -> paddingHorizontal
        padding > 0f -> padding
        else -> 0f
    }

    fun resolvedPaddingRight(): Float = when {
        paddingRight > 0f -> paddingRight
        paddingHorizontal > 0f -> paddingHorizontal
        padding > 0f -> padding
        else -> 0f
    }

    // --- Margin resolution (same cascade) ---

    fun resolvedMarginTop(): Float = when {
        marginTop > 0f -> marginTop
        marginVertical > 0f -> marginVertical
        margin > 0f -> margin
        else -> 0f
    }

    fun resolvedMarginBottom(): Float = when {
        marginBottom > 0f -> marginBottom
        marginVertical > 0f -> marginVertical
        margin > 0f -> margin
        else -> 0f
    }

    fun resolvedMarginLeft(): Float = when {
        marginLeft > 0f -> marginLeft
        marginHorizontal > 0f -> marginHorizontal
        margin > 0f -> margin
        else -> 0f
    }

    fun resolvedMarginRight(): Float = when {
        marginRight > 0f -> marginRight
        marginHorizontal > 0f -> marginHorizontal
        margin > 0f -> margin
        else -> 0f
    }

    // --- Border resolution (cascade: specific > general) ---

    fun resolvedBorderTopWidth(): Float =
        if (borderTopWidth > 0f) borderTopWidth else borderWidth

    fun resolvedBorderBottomWidth(): Float =
        if (borderBottomWidth > 0f) borderBottomWidth else borderWidth

    fun resolvedBorderLeftWidth(): Float =
        if (borderLeftWidth > 0f) borderLeftWidth else borderWidth

    fun resolvedBorderRightWidth(): Float =
        if (borderRightWidth > 0f) borderRightWidth else borderWidth

    fun resolvedBorderTopColor(): Int? =
        borderTopColor ?: borderColor

    fun resolvedBorderBottomColor(): Int? =
        borderBottomColor ?: borderColor

    fun resolvedBorderLeftColor(): Int? =
        borderLeftColor ?: borderColor

    fun resolvedBorderRightColor(): Int? =
        borderRightColor ?: borderColor

    fun hasAnyBorder(): Boolean =
        resolvedBorderTopWidth() > 0f || resolvedBorderBottomWidth() > 0f ||
            resolvedBorderLeftWidth() > 0f || resolvedBorderRightWidth() > 0f

    companion object {
        fun fromJSON(json: JSONObject?): ElementStyle {
            if (json == null) return ElementStyle()
            return ElementStyle(
                fontSize = json.optDouble("fontSize", 0.0).toFloat(),
                fontWeight = optStringOrNull(json, "fontWeight"),
                fontStyle = optStringOrNull(json, "fontStyle"),
                fontFamily = optStringOrNull(json, "fontFamily"),
                lineHeight = json.optDouble("lineHeight", 0.0).toFloat(),
                letterSpacing = json.optDouble("letterSpacing", 0.0).toFloat(),
                textDecorationLine = optStringOrNull(json, "textDecorationLine"),
                textDecorationColor = parseColor(json.opt("textDecorationColor")),
                textDecorationStyle = optStringOrNull(json, "textDecorationStyle"),
                textAlign = optStringOrNull(json, "textAlign"),
                color = parseColor(json.opt("color")),
                backgroundColor = parseColor(json.opt("backgroundColor")),
                gap = json.optDouble("gap", 0.0).toFloat(),
                width = json.optDouble("width", 0.0).toFloat(),
                height = json.optDouble("height", 0.0).toFloat(),
                maxWidth = json.optDouble("maxWidth", 0.0).toFloat(),
                maxHeight = json.optDouble("maxHeight", 0.0).toFloat(),
                padding = json.optDouble("padding", 0.0).toFloat(),
                paddingHorizontal = json.optDouble("paddingHorizontal", 0.0).toFloat(),
                paddingVertical = json.optDouble("paddingVertical", 0.0).toFloat(),
                paddingTop = json.optDouble("paddingTop", 0.0).toFloat(),
                paddingBottom = json.optDouble("paddingBottom", 0.0).toFloat(),
                paddingLeft = json.optDouble("paddingLeft", 0.0).toFloat(),
                paddingRight = json.optDouble("paddingRight", 0.0).toFloat(),
                margin = json.optDouble("margin", 0.0).toFloat(),
                marginHorizontal = json.optDouble("marginHorizontal", 0.0).toFloat(),
                marginVertical = json.optDouble("marginVertical", 0.0).toFloat(),
                marginTop = json.optDouble("marginTop", 0.0).toFloat(),
                marginBottom = json.optDouble("marginBottom", 0.0).toFloat(),
                marginLeft = json.optDouble("marginLeft", 0.0).toFloat(),
                marginRight = json.optDouble("marginRight", 0.0).toFloat(),
                borderColor = parseColor(json.opt("borderColor")),
                borderWidth = json.optDouble("borderWidth", 0.0).toFloat(),
                borderRadius = json.optDouble("borderRadius", 0.0).toFloat(),
                borderLeftColor = parseColor(json.opt("borderLeftColor")),
                borderLeftWidth = json.optDouble("borderLeftWidth", 0.0).toFloat(),
                borderRightColor = parseColor(json.opt("borderRightColor")),
                borderRightWidth = json.optDouble("borderRightWidth", 0.0).toFloat(),
                borderTopColor = parseColor(json.opt("borderTopColor")),
                borderTopWidth = json.optDouble("borderTopWidth", 0.0).toFloat(),
                borderBottomColor = parseColor(json.opt("borderBottomColor")),
                borderBottomWidth = json.optDouble("borderBottomWidth", 0.0).toFloat(),
            )
        }

        private fun optStringOrNull(json: JSONObject, key: String): String? {
            return if (json.has(key) && !json.isNull(key)) json.optString(key) else null
        }

        private fun parseColor(value: Any?): Int? {
            return when (value) {
                is Number -> value.toInt()
                is String -> try { Color.parseColor(value) } catch (_: Exception) { null }
                else -> null
            }
        }
    }
}

data class StyleConfig(
    /** Base text style — applies to all text unless overridden */
    val base: ElementStyle = ElementStyle(),
    val text: ElementStyle = ElementStyle(),
    val heading1: ElementStyle = ElementStyle(),
    val heading2: ElementStyle = ElementStyle(),
    val heading3: ElementStyle = ElementStyle(),
    val heading4: ElementStyle = ElementStyle(),
    val heading5: ElementStyle = ElementStyle(),
    val heading6: ElementStyle = ElementStyle(),
    val paragraph: ElementStyle = ElementStyle(),
    val strong: ElementStyle = ElementStyle(),
    val emphasis: ElementStyle = ElementStyle(),
    val strikethrough: ElementStyle = ElementStyle(),
    val code: ElementStyle = ElementStyle(),
    val codeBlock: ElementStyle = ElementStyle(),
    val link: ElementStyle = ElementStyle(),
    val blockquote: ElementStyle = ElementStyle(),
    val list: ElementStyle = ElementStyle(),
    val listItem: ElementStyle = ElementStyle(),
    val listBullet: ElementStyle = ElementStyle(),
    val table: ElementStyle = ElementStyle(),
    val tableRow: ElementStyle = ElementStyle(),
    val tableHeaderRow: ElementStyle = ElementStyle(),
    val tableCell: ElementStyle = ElementStyle(),
    val tableHeaderCell: ElementStyle = ElementStyle(),
    val thematicBreak: ElementStyle = ElementStyle(),
    val image: ElementStyle = ElementStyle(),
    val mentionUser: ElementStyle = ElementStyle(),
    val mentionChannel: ElementStyle = ElementStyle(),
    val mentionCommand: ElementStyle = ElementStyle(),
    val spoiler: ElementStyle = ElementStyle(),
) {
    fun styleForHeadingLevel(level: Int): ElementStyle = when (level) {
        1 -> heading1
        2 -> heading2
        3 -> heading3
        4 -> heading4
        5 -> heading5
        6 -> heading6
        else -> heading1
    }

    companion object {
        fun fromJSON(json: String?): StyleConfig {
            if (json.isNullOrEmpty()) return StyleConfig()
            return try {
                val obj = JSONObject(json)
                StyleConfig(
                    base = ElementStyle.fromJSON(obj.optJSONObject("base")),
                    text = ElementStyle.fromJSON(obj.optJSONObject("text")),
                    heading1 = ElementStyle.fromJSON(obj.optJSONObject("heading1")),
                    heading2 = ElementStyle.fromJSON(obj.optJSONObject("heading2")),
                    heading3 = ElementStyle.fromJSON(obj.optJSONObject("heading3")),
                    heading4 = ElementStyle.fromJSON(obj.optJSONObject("heading4")),
                    heading5 = ElementStyle.fromJSON(obj.optJSONObject("heading5")),
                    heading6 = ElementStyle.fromJSON(obj.optJSONObject("heading6")),
                    paragraph = ElementStyle.fromJSON(obj.optJSONObject("paragraph")),
                    strong = ElementStyle.fromJSON(obj.optJSONObject("strong")),
                    emphasis = ElementStyle.fromJSON(obj.optJSONObject("emphasis")),
                    strikethrough = ElementStyle.fromJSON(obj.optJSONObject("strikethrough")),
                    code = ElementStyle.fromJSON(obj.optJSONObject("code")),
                    codeBlock = ElementStyle.fromJSON(obj.optJSONObject("codeBlock")),
                    link = ElementStyle.fromJSON(obj.optJSONObject("link")),
                    blockquote = ElementStyle.fromJSON(obj.optJSONObject("blockquote")),
                    list = ElementStyle.fromJSON(obj.optJSONObject("list")),
                    listItem = ElementStyle.fromJSON(obj.optJSONObject("listItem")),
                    listBullet = ElementStyle.fromJSON(obj.optJSONObject("listBullet")),
                    table = ElementStyle.fromJSON(obj.optJSONObject("table")),
                    tableRow = ElementStyle.fromJSON(obj.optJSONObject("tableRow")),
                    tableHeaderRow = ElementStyle.fromJSON(obj.optJSONObject("tableHeaderRow")),
                    tableCell = ElementStyle.fromJSON(obj.optJSONObject("tableCell")),
                    tableHeaderCell = ElementStyle.fromJSON(obj.optJSONObject("tableHeaderCell")),
                    thematicBreak = ElementStyle.fromJSON(obj.optJSONObject("thematicBreak")),
                    image = ElementStyle.fromJSON(obj.optJSONObject("image")),
                    mentionUser = ElementStyle.fromJSON(obj.optJSONObject("mentionUser")),
                    mentionChannel = ElementStyle.fromJSON(obj.optJSONObject("mentionChannel")),
                    mentionCommand = ElementStyle.fromJSON(obj.optJSONObject("mentionCommand")),
                    spoiler = ElementStyle.fromJSON(obj.optJSONObject("spoiler")),
                )
            } catch (_: Exception) {
                StyleConfig()
            }
        }
    }
}

File: ios/styles/StyleConfig.m
#import "StyleConfig.h"

#import <React/RCTConvert.h>

@implementation MarkdownElementStyle

- (instancetype)init {
  self = [super init];
  if (self) {
    // Use NaN as sentinel for "not set" so that explicit zero values
    // are distinguishable from the default unset state.
    _fontSize = NAN;
    _letterSpacing = NAN;
    _lineHeight = NAN;
    _gap = NAN;
    _width = NAN;
    _height = NAN;
    _maxWidth = NAN;
    _maxHeight = NAN;
    _margin = NAN;
    _marginTop = NAN;
    _marginBottom = NAN;
    _marginLeft = NAN;
    _marginRight = NAN;
    _marginHorizontal = NAN;
    _marginVertical = NAN;
    _padding = NAN;
    _paddingTop = NAN;
    _paddingBottom = NAN;
    _paddingLeft = NAN;
    _paddingRight = NAN;
    _paddingHorizontal = NAN;
    _paddingVertical = NAN;
    _borderWidth = NAN;
    _borderTopWidth = NAN;
    _borderBottomWidth = NAN;
    _borderLeftWidth = NAN;
    _borderRightWidth = NAN;
    _borderRadius = NAN;
    _borderTopLeftRadius = NAN;
    _borderTopRightRadius = NAN;
    _borderBottomLeftRadius = NAN;
    _borderBottomRightRadius = NAN;
  }
  return self;
}

#pragma mark - Font resolution

- (UIFont *)resolvedFont {
  return [self resolvedFontWithBase:nil];
}

- (UIFont *)resolvedFontWithBase:(UIFont *)baseFont {
  CGFloat size = !isnan(_fontSize)
      ? _fontSize
      : (baseFont ? baseFont.pointSize : 0);
  if (size <= 0) return baseFont;

  NSString *family = _fontFamily ?: baseFont.familyName;

  // Start from base font's traits if available
  UIFontDescriptorSymbolicTraits traits =
      baseFont ? baseFont.fontDescriptor.symbolicTraits : 0;

  if (_fontWeight) {
    if ([_fontWeight isEqualToString:@"bold"] ||
        [_fontWeight isEqualToString:@"600"] ||
        [_fontWeight isEqualToString:@"700"] ||
        [_fontWeight isEqualToString:@"800"] ||
        [_fontWeight isEqualToString:@"900"]) {
      traits |= UIFontDescriptorTraitBold;
    } else if ([_fontWeight isEqualToString:@"normal"] ||
               [_fontWeight isEqualToString:@"100"] ||
               [_fontWeight isEqualToString:@"200"] ||
               [_fontWeight isEqualToString:@"300"] ||
               [_fontWeight isEqualToString:@"400"] ||
               [_fontWeight isEqualToString:@"500"]) {
      traits &= ~UIFontDescriptorTraitBold;
    }
  }

  if (_fontStyle) {
    if ([_fontStyle isEqualToString:@"italic"]) {
      traits |= UIFontDescriptorTraitItalic;
    } else if ([_fontStyle isEqualToString:@"normal"]) {
      traits &= ~UIFontDescriptorTraitItalic;
    }
  }

  UIFont *resolved = nil;
  if (!family || [family isEqualToString:@"System"]) {
    UIFontWeight weight = (traits & UIFontDescriptorTraitBold)
        ? UIFontWeightBold
        : UIFontWeightRegular;
    resolved = [UIFont systemFontOfSize:size weight:weight];
    if (traits & UIFontDescriptorTraitItalic) {
      UIFontDescriptor *d =
          [resolved.fontDescriptor fontDescriptorWithSymbolicTraits:
              resolved.fontDescriptor.symbolicTraits | UIFontDescriptorTraitItalic];
      if (d) resolved = [UIFont fontWithDescriptor:d size:size];
    }
  } else {
    UIFont *base = [UIFont fontWithName:family size:size];
    if (base) {
      UIFontDescriptor *d =
          [base.fontDescriptor fontDescriptorWithSymbolicTraits:traits];
      resolved = d ? [UIFont fontWithDescriptor:d size:size] : base;
    }
  }

  return resolved ?: baseFont;
}

#pragma mark - Layout insets

- (UIEdgeInsets)resolvedPaddingInsets {
  // Specific edges override paddingHorizontal/paddingVertical which override padding.
  // NaN means "not set" so explicit 0 values are respected.
  CGFloat base = !isnan(_padding) ? _padding : 0;
  CGFloat vBase = !isnan(_paddingVertical) ? _paddingVertical : base;
  CGFloat hBase = !isnan(_paddingHorizontal) ? _paddingHorizontal : base;
  CGFloat top = !isnan(_paddingTop) ? _paddingTop : vBase;
  CGFloat bottom = !isnan(_paddingBottom) ? _paddingBottom : vBase;
  CGFloat left = !isnan(_paddingLeft) ? _paddingLeft : hBase;
  CGFloat right = !isnan(_paddingRight) ? _paddingRight : hBase;

  return UIEdgeInsetsMake(top, left, bottom, right);
}

- (UIEdgeInsets)resolvedMarginInsets {
  CGFloat base = !isnan(_margin) ? _margin : 0;
  CGFloat vBase = !isnan(_marginVertical) ? _marginVertical : base;
  CGFloat hBase = !isnan(_marginHorizontal) ? _marginHorizontal : base;
  CGFloat top = !isnan(_marginTop) ? _marginTop : vBase;
  CGFloat bottom = !isnan(_marginBottom) ? _marginBottom : vBase;
  CGFloat left = !isnan(_marginLeft) ? _marginLeft : hBase;
  CGFloat right = !isnan(_marginRight) ? _marginRight : hBase;

  return UIEdgeInsetsMake(top, left, bottom, right);
}

- (UIEdgeInsets)resolvedBorderWidths {
  CGFloat base = !isnan(_borderWidth) ? _borderWidth : 0;
  CGFloat top = !isnan(_borderTopWidth) ? _borderTopWidth : base;
  CGFloat bottom = !isnan(_borderBottomWidth) ? _borderBottomWidth : base;
  CGFloat left = !isnan(_borderLeftWidth) ? _borderLeftWidth : base;
  CGFloat right = !isnan(_borderRightWidth) ? _borderRightWidth : base;
  return UIEdgeInsetsMake(top, left, bottom, right);
}

- (UIColor *)resolvedBorderColorForEdge:(UIRectEdge)edge {
  switch (edge) {
    case UIRectEdgeTop:
      return _borderTopColor ?: _borderColor;
    case UIRectEdgeBottom:
      return _borderBottomColor ?: _borderColor;
    case UIRectEdgeLeft:
      return _borderLeftColor ?: _borderColor;
    case UIRectEdgeRight:
      return _borderRightColor ?: _borderColor;
    default:
      return _borderColor;
  }
}

- (CGFloat)resolvedRadiusForCorner:(UIRectCorner)corner {
  CGFloat base = !isnan(_borderRadius) ? _borderRadius : 0;
  switch (corner) {
    case UIRectCornerTopLeft:
      return !isnan(_borderTopLeftRadius) ? _borderTopLeftRadius : base;
    case UIRectCornerTopRight:
      return !isnan(_borderTopRightRadius) ? _borderTopRightRadius : base;
    case UIRectCornerBottomLeft:
      return !isnan(_borderBottomLeftRadius) ? _borderBottomLeftRadius : base;
    case UIRectCornerBottomRight:
      return !isnan(_borderBottomRightRadius) ? _borderBottomRightRadius : base;
    default:
      return base;
  }
}

- (BOOL)hasAnyBorder {
  UIEdgeInsets widths = [self resolvedBorderWidths];
  return widths.top > 0 || widths.bottom > 0 || widths.left > 0 || widths.right > 0;
}

- (BOOL)hasAnyRadius {
  return (!isnan(_borderRadius) && _borderRadius > 0) ||
         (!isnan(_borderTopLeftRadius) && _borderTopLeftRadius > 0) ||
         (!isnan(_borderTopRightRadius) && _borderTopRightRadius > 0) ||
         (!isnan(_borderBottomLeftRadius) && _borderBottomLeftRadius > 0) ||
         (!isnan(_borderBottomRightRadius) && _borderBottomRightRadius > 0);
}

- (BOOL)hasNonUniformBorders {
  UIEdgeInsets widths = [self resolvedBorderWidths];
  BOOL widthsDiffer = !(widths.top == widths.bottom &&
                        widths.bottom == widths.left &&
                        widths.left == widths.right);
  if (widthsDiffer) return YES;

  UIColor *top = [self resolvedBorderColorForEdge:UIRectEdgeTop];
  UIColor *bottom = [self resolvedBorderColorForEdge:UIRectEdgeBottom];
  UIColor *left = [self resolvedBorderColorForEdge:UIRectEdgeLeft];
  UIColor *right = [self resolvedBorderColorForEdge:UIRectEdgeRight];

  if ((top && bottom && ![top isEqual:bottom]) ||
      (top && left && ![top isEqual:left]) ||
      (top && right && ![top isEqual:right])) {
    return YES;
  }

  return NO;
}

@end

#pragma mark - StyleConfig

@implementation StyleConfig

+ (instancetype)fromJSON:(NSString *)json {
  StyleConfig *config = [[StyleConfig alloc] init];
  if (!json || json.length == 0) return config;

  NSData *data = [json dataUsingEncoding:NSUTF8StringEncoding];
  NSError *error;
  NSDictionary *dict = [NSJSONSerialization JSONObjectWithData:data
                                                       options:0
                                                         error:&error];
  if (error || ![dict isKindOfClass:[NSDictionary class]]) return config;

  config.base = [self elementStyleFromDict:dict[@"base"]];

  config.paragraph = [self elementStyleFromDict:dict[@"paragraph"]];
  config.heading1 = [self elementStyleFromDict:dict[@"heading1"]];
  config.heading2 = [self elementStyleFromDict:dict[@"heading2"]];
  config.heading3 = [self elementStyleFromDict:dict[@"heading3"]];
  config.heading4 = [self elementStyleFromDict:dict[@"heading4"]];
  config.heading5 = [self elementStyleFromDict:dict[@"heading5"]];
  config.heading6 = [self elementStyleFromDict:dict[@"heading6"]];
  config.blockquote = [self elementStyleFromDict:dict[@"blockquote"]];
  config.codeBlock = [self elementStyleFromDict:dict[@"codeBlock"]];
  config.list = [self elementStyleFromDict:dict[@"list"]];
  config.listItem = [self elementStyleFromDict:dict[@"listItem"]];
  config.listBullet = [self elementStyleFromDict:dict[@"listBullet"]];
  config.thematicBreak = [self elementStyleFromDict:dict[@"thematicBreak"]];
  config.image = [self elementStyleFromDict:dict[@"image"]];

  config.table = [self elementStyleFromDict:dict[@"table"]];
  config.tableRow = [self elementStyleFromDict:dict[@"tableRow"]];
  config.tableHeaderRow = [self elementStyleFromDict:dict[@"tableHeaderRow"]];
  config.tableCell = [self elementStyleFromDict:dict[@"tableCell"]];
  config.tableHeaderCell = [self elementStyleFromDict:dict[@"tableHeaderCell"]];

  config.strong = [self elementStyleFromDict:dict[@"strong"]];
  config.emphasis = [self elementStyleFromDict:dict[@"emphasis"]];
  config.strikethrough = [self elementStyleFromDict:dict[@"strikethrough"]];
  config.code = [self elementStyleFromDict:dict[@"code"]];
  config.link = [self elementStyleFromDict:dict[@"link"]];
  config.mentionUser = [self elementStyleFromDict:dict[@"mentionUser"]];
  config.mentionChannel = [self elementStyleFromDict:dict[@"mentionChannel"]];
  config.mentionCommand = [self elementStyleFromDict:dict[@"mentionCommand"]];
  config.spoiler = [self elementStyleFromDict:dict[@"spoiler"]];
  config.superscript = [self elementStyleFromDict:dict[@"superscript"]];

  return config;
}

+ (MarkdownElementStyle *)elementStyleFromDict:(NSDictionary *)dict {
  MarkdownElementStyle *style = [[MarkdownElementStyle alloc] init];
  if (![dict isKindOfClass:[NSDictionary class]]) return style;

  // Text properties
  style.color = [self colorFromValue:dict[@"color"]];
  if (dict[@"fontFamily"]) style.fontFamily = dict[@"fontFamily"];
  if (dict[@"fontSize"]) style.fontSize = [dict[@"fontSize"] doubleValue];
  if (dict[@"fontStyle"]) style.fontStyle = dict[@"fontStyle"];
  if (dict[@"fontWeight"]) style.fontWeight = dict[@"fontWeight"];
  if (dict[@"letterSpacing"]) style.letterSpacing = [dict[@"letterSpacing"] doubleValue];
  if (dict[@"lineHeight"]) style.lineHeight = [dict[@"lineHeight"] doubleValue];
  if (dict[@"textAlign"]) style.textAlign = dict[@"textAlign"];
  style.textDecorationColor = [self colorFromValue:dict[@"textDecorationColor"]];
  if (dict[@"textDecorationLine"]) style.textDecorationLine = dict[@"textDecorationLine"];
  if (dict[@"textDecorationStyle"]) style.textDecorationStyle = dict[@"textDecorationStyle"];

  // View properties
  style.backgroundColor = [self colorFromValue:dict[@"backgroundColor"]];

  if (dict[@"gap"]) style.gap = [dict[@"gap"] doubleValue];
  if (dict[@"width"]) style.width = [dict[@"width"] doubleValue];
  if (dict[@"height"]) style.height = [dict[@"height"] doubleValue];
  if (dict[@"maxWidth"]) style.maxWidth = [dict[@"maxWidth"] doubleValue];
  if (dict[@"maxHeight"]) style.maxHeight = [dict[@"maxHeight"] doubleValue];
  if (dict[@"objectFit"]) style.objectFit = dict[@"objectFit"];

  // Margin
  if (dict[@"margin"]) style.margin = [dict[@"margin"] doubleValue];
  if (dict[@"marginTop"]) style.marginTop = [dict[@"marginTop"] doubleValue];
  if (dict[@"marginBottom"]) style.marginBottom = [dict[@"marginBottom"] doubleValue];
  if (dict[@"marginLeft"]) style.marginLeft = [dict[@"marginLeft"] doubleValue];
  if (dict[@"marginRight"]) style.marginRight = [dict[@"marginRight"] doubleValue];
  if (dict[@"marginHorizontal"]) style.marginHorizontal = [dict[@"marginHorizontal"] doubleValue];
  if (dict[@"marginVertical"]) style.marginVertical = [dict[@"marginVertical"] doubleValue];

  // Padding
  if (dict[@"padding"]) style.padding = [dict[@"padding"] doubleValue];
  if (dict[@"paddingTop"]) style.paddingTop = [dict[@"paddingTop"] doubleValue];
  if (dict[@"paddingBottom"]) style.paddingBottom = [dict[@"paddingBottom"] doubleValue];
  if (dict[@"paddingLeft"]) style.paddingLeft = [dict[@"paddingLeft"] doubleValue];
  if (dict[@"paddingRight"]) style.paddingRight = [dict[@"paddingRight"] doubleValue];
  if (dict[@"paddingHorizontal"]) style.paddingHorizontal = [dict[@"paddingHorizontal"] doubleValue];
  if (dict[@"paddingVertical"]) style.paddingVertical = [dict[@"paddingVertical"] doubleValue];

  // Border widths
  if (dict[@"borderWidth"]) style.borderWidth = [dict[@"borderWidth"] doubleValue];
  if (dict[@"borderTopWidth"]) style.borderTopWidth = [dict[@"borderTopWidth"] doubleValue];
  if (dict[@"borderBottomWidth"]) style.borderBottomWidth = [dict[@"borderBottomWidth"] doubleValue];
  if (dict[@"borderLeftWidth"]) style.borderLeftWidth = [dict[@"borderLeftWidth"] doubleValue];
  if (dict[@"borderRightWidth"]) style.borderRightWidth = [dict[@"borderRightWidth"] doubleValue];

  // Border colors
  style.borderColor = [self colorFromValue:dict[@"borderColor"]];
  style.borderTopColor = [self colorFromValue:dict[@"borderTopColor"]];
  style.borderBottomColor = [self colorFromValue:dict[@"borderBottomColor"]];
  style.borderLeftColor = [self colorFromValue:dict[@"borderLeftColor"]];
  style.borderRightColor = [self colorFromValue:dict[@"borderRightColor"]];

  // Border radii
  if (dict[@"borderRadius"]) style.borderRadius = [dict[@"borderRadius"] doubleValue];
  if (dict[@"borderTopLeftRadius"]) style.borderTopLeftRadius = [dict[@"borderTopLeftRadius"] doubleValue];
  if (dict[@"borderTopRightRadius"]) style.borderTopRightRadius = [dict[@"borderTopRightRadius"] doubleValue];
  if (dict[@"borderBottomLeftRadius"]) style.borderBottomLeftRadius = [dict[@"borderBottomLeftRadius"] doubleValue];
  if (dict[@"borderBottomRightRadius"]) style.borderBottomRightRadius = [dict[@"borderBottomRightRadius"] doubleValue];

  if (dict[@"borderStyle"]) style.borderStyle = dict[@"borderStyle"];
  if (dict[@"borderCurve"]) style.borderCurve = dict[@"borderCurve"];

  return style;
}

+ (UIColor *)colorFromValue:(id)value {
  if (!value || [value isKindOfClass:[NSNull class]]) return nil;

  // Delegate to React Native's own color converter. It handles
  // every shape processColor can emit:
  //   - NSNumber  : argb int from hex / rgb() / named strings
  //   - NSString  : hex / named / rgb(a) / hsl(a) fallback
  //   - NSDictionary with `semantic` key : PlatformColor('name')
  //     resolves via `colorNamed:` and iOS system color accessors
  //     like +labelColor, +systemRedColor, etc.
  //   - NSDictionary with `dynamic` key  : DynamicColorIOS({ light,
  //     dark, … }) becomes a UIColor that resolves per trait
  //     collection at draw time.
  // Rolling our own hex parser here missed both PlatformColor
  // dicts and anything beyond 6-char hex.
  return [RCTConvert UIColor:value];
}

- (MarkdownElementStyle *)styleForHeadingLevel:(NSInteger)level {
  switch (level) {
    case 1: return self.heading1;
    case 2: return self.heading2;
    case 3: return self.heading3;
    case 4: return self.heading4;
    case 5: return self.heading5;
    case 6: return self.heading6;
    default: return self.heading1;
  }
}

@end
